File: Fitness/ViewControllers/EditExerciseTableViewController.m
//
//  EditExerciseTableViewController.m
//  Fitness
//
//  Created by Mani on 23.04.15.
//  Copyright (c) 2015 Mani. All rights reserved.
//

#import "EditExerciseTableViewController.h"

@interface EditExerciseTableViewController () <UIPickerViewDataSource, UIPickerViewDelegate>
{
    // Initial data
    enum ExerciseType _initialExerciseType;
    NSTimeInterval _initialWarmup;
    NSTimeInterval _initialInterval;
}

@property (strong, nonatomic) IBOutlet UIPickerView *exercisePickerView;
@property (strong, nonatomic) IBOutlet UIPickerView *warmUpPickerView;
@property (strong, nonatomic) IBOutlet UIPickerView *intervalPickerView;

@end

@implementation EditExerciseTableViewController

- (IBAction)onSave:(id)sender {
    
    enum ExerciseType type = (enum ExerciseType)[self.exercisePickerView selectedRowInComponent:0];
    NSTimeInterval warmup = (NSTimeInterval)([self.warmUpPickerView selectedRowInComponent:0] + 5);
    NSTimeInterval interval = (NSTimeInterval)([self.intervalPickerView selectedRowInComponent:0] * 60 + [self.intervalPickerView selectedRowInComponent:1]);
    
    if(_delegate != nil)
        [_delegate editExerciseTableViewController:self didFinishedWithExerciseType:type withWarmupInterval:warmup withTimeInterval:interval];
}


- (IBAction)onCancel:(id)sender {
    if(_delegate != nil)
        [_delegate editExerciseTableViewControllerDidCancel:self];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _initialExerciseType = 0;
    _initialWarmup = 5;
    _initialInterval = 60;
    
    self.exercisePickerView.delegate = self;
    self.warmUpPickerView.delegate = self;
    self.intervalPickerView.delegate = self;
    
    self.exercisePickerView.dataSource = self;
    self.warmUpPickerView.dataSource = self;
    self.intervalPickerView.dataSource = self;
    
    [self.exercisePickerView selectRow:_initialExerciseType inComponent:0 animated:NO];
    [self.warmUpPickerView selectRow:_initialWarmup - 5 inComponent:0 animated:NO];
    [self.intervalPickerView selectRow:(NSInteger)((NSInteger)(_initialInterval) / 60) inComponent:0 animated:NO];
    [self.intervalPickerView selectRow:(NSInteger)((NSInteger)(_initialInterval) % 60) inComponent:1 animated:NO];
    
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    // The interval picker also supports minutes, so we need two components
    return (pickerView == self.intervalPickerView) ? 2 : 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    
    if(pickerView == self.exercisePickerView)
        return (NSInteger)ExerciseTypeCount;
    
    if(pickerView == self.warmUpPickerView)
        return 16;
    
    if(pickerView == self.intervalPickerView) {
        if(component == 0) return 120; // 120 minutes
        else return 60; // 60 seconds
    }
    
    return 0;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    if(pickerView == self.exercisePickerView)
        return [CDExercise exerciseTypeToString:(enum ExerciseType)row];
    
    if(pickerView == self.warmUpPickerView)
        return [NSString stringWithFormat:@"%ld Sekunden", row + 5];
    
    if(pickerView == self.intervalPickerView) {
        if(component == 0)  // minutes
            return [NSString stringWithFormat:@"%ld Minuten", row];
        else // seconds
            return [NSString stringWithFormat:@"%ld Sekunden", row];
    }
    
    return @"";
}

- (BOOL)setInitialData:(enum ExerciseType)type warmupTime:(NSTimeInterval)warmup timeInterval:(NSTimeInterval)interval {
    _initialExerciseType = type;
    _initialWarmup = warmup;
    _initialInterval = interval;
    
    return YES;
}

@end

File: Fitness/ViewControllers/EditExerciseTableViewController.swift
//
//  EditExerciseTableViewController.swift
//  Fitness
//
//  Created by Manuel Stampfl on 25.06.15.
//  Copyright (c) 2015 mani1337. All rights reserved.
//

import UIKit

@objc protocol EditExerciseTableViewControllerDelegate {
    func editExerciseTableViewControllerDidCancelled(controller: EditExerciseTableViewController)
    func editExerciseTableViewControllerDidFinished(controller: EditExerciseTableViewController)
}

class EditExerciseTableViewController: UITableViewController, UIPickerViewDataSource, UIPickerViewDelegate {
    // MARK: - Vars
    weak var delegate: EditExerciseTableViewControllerDelegate?
    var exercise:Exercise?
    
    // MARK: - Storyboard Outlets
    @IBOutlet weak var typePickerView: UIPickerView!
    @IBOutlet weak var warmupPickerView: UIPickerView!
    @IBOutlet weak var durationTimerView: UIPickerView!
    
    // MARK: - Storyboard Actions
    @IBAction func onBarButtonItemSave(sender: AnyObject) {
        if self.exercise == nil {
            self.exercise = Exercise()
        }
        
        self.exercise!.type = ExerciseType(rawValue: self.typePickerView.selectedRowInComponent(0))!
        self.exercise!.warmup = NSTimeInterval(self.warmupPickerView.selectedRowInComponent(0) + 5)
        self.exercise!.duration = NSTimeInterval(self.durationTimerView.selectedRowInComponent(0) + 1)
        
        self.delegate?.editExerciseTableViewControllerDidFinished(self)
    }
    
    @IBAction func onBarButtonItemClose(sender: AnyObject) {
        delegate?.editExerciseTableViewControllerDidCancelled(self)
    }
    
    // MARK: - Overrided Base Methods
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.typePickerView.delegate = self
        self.typePickerView.dataSource = self
        
        self.warmupPickerView.delegate = self
        self.warmupPickerView.dataSource = self
        
        self.durationTimerView.delegate = self
        self.durationTimerView.dataSource = self
        
        var type: ExerciseType = .Running
        var warmup: NSTimeInterval = 5.0
        var duration: NSTimeInterval = 60.0
        if self.exercise != nil {
            type = self.exercise!.type
            warmup = self.exercise!.warmup
            duration = self.exercise!.duration
        }
        
        self.typePickerView.selectRow(type.rawValue, inComponent: 0, animated: false)
        self.warmupPickerView.selectRow(Int(warmup) - 5, inComponent: 0, animated: false)
        self.durationTimerView.selectRow(Int(duration) - 1, inComponent: 0, animated: false)
    }
    

    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        switch pickerView {
        case self.typePickerView:
            return ExerciseType.Count.rawValue
        case self.warmupPickerView:
            return 11 // 5 - 15 seconds
        case self.durationTimerView:
            return 3600
        default:
            return 0
        }
    }
    
    func pickerView(pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String! {
        switch pickerView {
        case self.typePickerView:
            return ExerciseType(rawValue: row)!.localizedName()
        case self.warmupPickerView:
            return "\(row + 5) " + NSLocalizedString("SECONDS", comment: "seconds")
        case self.durationTimerView:
            return "\(row + 1) " + NSLocalizedString("SECONDS", comment: "seconds")
        default:
            return ""
        }
    }
}
